src/net/download.rs

```diff
@@ -0,0 +1,1904 @@
+                    downloader.getmicroblocks_try_finish(http)
+                },
+                None => {
+                    test_debug!("{:?}: HTTP not connected", &network.local_peer);
+                    Err(net_error::NotConnected)
+                }
+            }
+        })
+    }
+
+    /// Process newly-fetched blocks and microblocks.
+    /// Returns true if we've completed all requests.
+    /// Returns (done?, blocks-we-got, microblocks-we-got) on success
+    fn finish_downloads(&mut self, burndb: &mut BurnDB, chainstate: &mut StacksChainState) -> Result<(bool, Vec<StacksBlock>, Vec<Vec<StacksMicroblock>>), net_error> {
+        let mut blocks = vec![];
+        let mut microblocks = vec![];
+        let mut done = false;
+
+        PeerNetwork::with_downloader_state(self, |ref mut network, ref mut downloader| {
+            // extract blocks and microblocks downloaded
+            for (request_key, block) in downloader.blocks.drain() {
+                test_debug!("Downloaded block {}/{} ({}) at sortition height {}", &request_key.burn_block_hash, &request_key.anchor_block_hash, &request_key.index_block_hash, request_key.sortition_height);
+                blocks.push(block);
+                downloader.num_blocks_downloaded += 1;
+
+                // don't try this again
+                downloader.blocks_to_try.remove(&request_key.sortition_height);
+            }
+            for (request_key, microblock_stream) in downloader.microblocks.drain() {
+                let block_header = StacksChainState::load_block_header(&chainstate.blocks_path, &request_key.burn_block_hash, &request_key.anchor_block_hash)? 
+                    .expect(&format!("BUG: missing Stacks block header for {}/{}", &request_key.burn_block_hash, &request_key.anchor_block_hash));
+
+                assert!(request_key.child_block_header.is_some(), "BUG: requested a microblock but didn't set the child block header");
+                let child_block_header = request_key.child_block_header.unwrap();
+
+                if StacksChainState::validate_parent_microblock_stream(&block_header, &child_block_header, &microblock_stream, true).is_some() {
+                    // stream is valid!
+                    test_debug!("Downloaded valid microblock stream {}/{} at sortition height {}", &request_key.burn_block_hash, &request_key.anchor_block_hash, request_key.sortition_height);
+                    microblocks.push(microblock_stream);
+                    downloader.num_microblocks_downloaded += 1;
+                }
+                else {
+                    // stream is not well-formed
+                    test_debug!("Microblock stream {:?}: {}/{} is invalid", request_key.sortition_height, &request_key.burn_block_hash, &request_key.anchor_block_hash);
+                }
+
+                // don't try again
+                downloader.microblocks_to_try.remove(&request_key.sortition_height);
+            }
+
+            // clear empties
```

Comment:
You can use HashMap::retain for patterns like iterate-and-remove: https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.retain